src/Http/Paginators/RequestPaginator.php

```diff
@@ -0,0 +1,281 @@
+        $this->async();
+
+        return $this->connector->pool(
+            $this,
+            $concurrency,
+            $responseHandler,
+            $exceptionHandler,
+        );
+    }
+
+    /**
+     * @param bool $async
+     *
+     * @return $this
+     */
+    public function async(bool $async = true): static
+    {
+        $this->async = $async;
+
+        return $this;
+    }
+
+    /**
+     * @return bool
+     */
+    public function isAsync(): bool
+    {
+        return $this->async;
+    }
+
+    /**
+     * @param bool $ignoreRewinding
+     *
+     * @return $this
+     */
+    public function ignoreRewinding(bool $ignoreRewinding = true): static
+    {
+        $this->shouldRewind = ! $ignoreRewinding;
+
+        return $this;
+    }
+
+    /**
+     * @return bool
+     */
+    public function shouldRewind(): bool
+    {
+        return $this->shouldRewind;
+    }
+
+    /**
+     * @return int
+     */
+    public function count(): int
+    {
+        return $this->totalPages();
+    }
+
+    /**
+     * @return \Saloon\Contracts\Response|null
+     */
+    public function currentResponse(): ?Response
+    {
+        return $this->currentResponse;
+    }
+
+    /**
+     * @return int|null
+     */
+    public function limit(): ?int
+    {
+        return $this->limit;
+    }
+
+    /**
+     * @return bool
+     *
+     * @TODO: Make naming more abstract and versatile.
+     *        F.e., page, offset
+     */
+    public function hasPreviousPage(): bool
+    {
+        return ! is_null($this->previousPage());
+    }
+
+    /**
+     * @return bool
+     *
+     * @TODO: Make naming more abstract and versatile.
+     *        F.e., page, offset
+     */
+    public function hasNextPage(): bool
+    {
+        return ! is_null($this->nextPage());
+    }
+
+    /**
+     * @return void
+     */
+    public function rewind(): void
+    {
+        if (! $this->shouldRewind()) {
+            return;
+        }
+
+        // Nullify the current response, so we don't accidentally check if there are more pages, even though we're starting over.
+        $this->currentResponse = null;
+    }
+
+    /**
+     * @return bool
+     */
+    public function valid(): bool
+    {
+        // If we haven't sent a request yet, and therefore don't have any response, the iterator is still valid.
+        // It's only ever invalid when we have sent a request, retrieved the response, and have no more pages after that.
+        // Otherwise PHP would immediately terminate  the iteration, without sending a request.
+        if (\is_null($this->currentResponse)) {
+            return true;
+        }
+
+        return $this->hasNextPage();
+    }
+
+    /**
+     * @return ($this->async is true ? \GuzzleHttp\Promise\PromiseInterface : TResponse)
+     */
+    public function current(): Response|PromiseInterface
+    {
+        $this->applyPaging(
+            $request = clone $this->originalRequest,
+        );
```

Comment:
This is clever!

Reply:
Still think modifying the PendingRequest might be better? I can't decide

Reply:
Strictly speaking, we want to send a new request.
Which is why I want to follow the regular sending-flow.

I'm not sure what the difference would be of creating a PendingRequest instead, as we'd still clone the initial Request.

Reply:
Okay, happy with that now :D